bin/bugsnag-maze-runner

```diff
@@ -142,20 +153,32 @@ class MazeRunnerEntry
     when :bs then
       raise "option --#{OPTION_USERNAME} must be specified" if options[OPTION_USERNAME].nil?
       raise "option --#{OPTION_ACCESS_KEY} must be specified" if options[OPTION_ACCESS_KEY].nil?
-      unless Devices::DEVICE_HASH.key? device_type
-        raise "Device type '#{device}' not known on BrowserStack.  Must be one of #{Devices::DEVICE_HASH.keys}"
+
+      bs_device = options[OPTION_BS_DEVICE]
+      raise "option --#{OPTION_BS_DEVICE} must be specified" if bs_device.nil?
+      unless Devices::DEVICE_HASH.key? bs_device
+        raise "Device type '#{bs_device}' not known on BrowserStack.  Must be one of #{Devices::DEVICE_HASH.keys}"
       end
 
+      config.bs_device = bs_device
+      config.os_version = Devices::DEVICE_HASH[config.bs_device]['os_version'].to_f
       config.bs_local = options[OPTION_BS_LOCAL]
       username = config.username = options[OPTION_USERNAME]
       access_key = config.access_key = options[OPTION_ACCESS_KEY]
       config.appium_server_url = "http://#{username}:#{access_key}@hub-cloud.browserstack.com/wd/hub"
     when :local then
-      config.appium_server_url = options[OPTION_APPIUM_SERVER]
-      device_type = device_type.downcase
-      raise 'device must be ios or android' unless %w[ios android].include? device_type
+      raise "option --#{OPTION_OS} must be specified" if options[OPTION_OS].nil?
+      raise "option --#{OPTION_OS_VERSION} must be specified" if options[OPTION_OS_VERSION].nil?
+      unless /^[1-9][0-9]*(\.[0-9])?/.match? options[OPTION_OS_VERSION]
```

Comment:
Ah, I've kept this in my head a bit, but yes it will become important when we start moving pipelines across to use this.  The Android pipeline, for example, contains a number of `skip` annotations, such as:
```
Before('@skip_above_android_7') do |scenario|
  skip_this_scenario("Skipping scenario") if %w[ANDROID_8_0 ANDROID_8_1 ANDROID_9_0 ANDROID_10_0].include? bs_device
end
```
Which has a number of drawbacks as it is (only works on BrowserStack, it's verbose and needs regular maintenance).  
Knowing that the OS version is a valid float and putting it into config will allow the above to collapse to just:
```
Before('@skip_above_android_7') do |scenario|
  skip_this_scenario("Skipping scenario") if MazeRunner.config.os_version > 7
end
```

Reply:
...I'll add a comment to try and capture that.

Reply:
Ruby has a class for version numbers in the stdlib which might be useful — https://ruby-doc.org/stdlib-2.7.2/libdoc/rubygems/rdoc/Gem/Version.html

It would make the comparison a bit more verbose, but allows for more formats of version numbers and it does the validation internally

```ruby
Before('@skip_above_android_7') do |scenario|
  skip_this_scenario("Skipping scenario") if MazeRunner.config.os_version > Gem::Version.create('7.0.0')
end
```

I wonder if we should provide a helper method in Maze Runner if it's a common thing needed across notifiers, especially as the OS version is now something it knows about, e.g.

```ruby
Before('@skip_above_android_7') do |scenario|
  MazeRunner.require_os_version('7.0.0')
end

// maze_runner.rb
class MazeRunner
  def self.require_os_version(version)
    skip_this_scenario if self.config.os_version > Gem::Version.create(version)
  end
end
```

or even add a set of skip annotations, so notifiers don't need to add them at all (with some logic to avoid adding duplicates)

```ruby
Devices.DEVICE_HASH.each do |device|
  Before("@skip_os_version_#{device[:os_version]}") do
    skip_this_scenario if MazeRunner.config.os_version == Gem::Version.create(device[:os_version])
  end

  Before("@skip_os_version_above_#{device[:os_version]}") do
    skip_this_scenario if MazeRunner.config.os_version > Gem::Version.create(device[:os_version])
  end

  Before("@skip_os_version_below_#{device[:os_version]}") do
    skip_this_scenario if MazeRunner.config.os_version < Gem::Version.create(device[:os_version])
  end
end
```

Reply:
As just discussed IRL, there's certainly scope for rolling more into MazeRunner in line with your suggestions.  At the moment I'm not totally sure how flexible the mechanism would need to be and if there are factors other than OS version that we would want to cater for.  So this is really only a small step along an exploratory trail, which will become clearer as we start to migrate notifier pipelines across to using it.  That said, this change will still simplify what notifiers need to do in their `env.rb` files.